### ISSUE TITLE
[BUG]: Audio player stops with internal "Unhandled Exception: killed by cleanCompleters"

### ISSUE THREAD

**Issue by heknt** [Canardoux/flutter_sound#876]
## Flutter Sound Version : 

- both **FULL** and **LITE** flavors
- Result of the command : ```flutter pub deps | grep flutter_sound```
> |-- flutter_sound_lite 8.5.0
> |   |-- flutter_sound_platform_interface 8.5.0
> |   |-- flutter_sound_web 8.5.0
> |   |   |-- flutter_sound_platform_interface...

----------------------------------------------------------

## Severity

- Result is not what expected
- App stucks(not crashes) with an internal exception that I can NOT catch

--------------------------------------------------------

## Platforms you faced the error 

- Android
- Real device

  We have got the reviews about the issue from this devices:
  - Samsung Galaxy A30 (Android 11)
  - Redmi Note 8 Pro (Android 11)
  - Huawei P10 Lite (Android 8)
  - Huawei P20 Lite (Android 9)
  - Huawei Honor 9 (Android 9)

------------------------------------------------

### Describe the bug
Our app ([link to the Play Market](https://play.google.com/store/apps/details?id=com.plus1s.neya)) has an Pronunciation exercise. Steps of this exercise: 
1. plays narrator's audio
2. records user's pronunciation
3. plays user's pronunciation
4. plays narrator's audio

There is a probability that program will stuck between 1st and 2nd points. Then we have no catched exception and program just idle at that screen.

We have a device (Huawei Honor 9 (Android 9)) which produces this exception during 3-10 minutes with 100% probability.
Honestly, at this device exercise is lagging with increasing over time till the end exception.

My device is Pixel 4a (Android 12). There was the cases when I produce this exception too. But for now it is an irrelevant for my device.

### To Reproduce
I have [a demo code](https://github.com/heknt/flutter_sound_recorder_usage) which has not produce the exception but still if you need it for test.

Steps to reproduce the behavior in the [our app from Play Market](https://play.google.com/store/apps/details?id=com.plus1s.neya)
1. Download an app
2. Move to the main screen
3. Press "Settings" button at the navigation bar
4. Press "Lesson Settings"
5. Choose only "Pronunciation" for a test
6. Move to the main screen
7. Press on the topic
8. Here we have a probability to get the issue after the first step from the paragraph "Describe the bug" - *plays narrator's audio*
----------------------------------------------

# Logs
Part of an exercise logs till the exception:
[error_at_the_end.log](https://github.com/Canardoux/flutter_sound/files/8468125/error_at_the_end.log)
Correct full logs of a successfully passed excercise:
[correct_full.log](https://github.com/Canardoux/flutter_sound/files/8468123/correct_full.log)

-----------------------------------------------------------------

**Comment by Larpoux** [comment]
The big problem is that the UI Widget is very buggy but is not maintained.
@Alvarocda had worked on a replacement, but unfortunately he has not yet be able to do a Pull Request.

**Comment by heknt** [comment]
@Larpoux What do you mean by UI Widget? Flutter Sound Player has a default UI representation? So I just use this class as a toggle of an audio playing with my own widgets.

**Comment by heknt** [comment]
@Larpoux I will wait for the @Alvarocda pull request.

About this issue,
`} on Error catch(e) {`
additionally to the existing Exception catching, helped me to avoid exercise blocking.

But generally problem still exists. You can keep this issue for the next update

**Comment by Larpoux** [comment]
Hi @heknt, my friend from Kiev

I changed the banner for : 
- [pub.dev](https://pub.dev/packages/flutter_sound)
- [github](https://github.com/canardoux/flutter_sound)
- [canardoux](https://flutter-sound.canardoux.xyz/readme.html)

Maybe you will have to refresh your pages in your browser if they are cached.
I like Street art. And I do not know what can I do more to help my Ukrainian brothers and sisters.
I am 70 years old and do not speak Russian or Ukrainian. And in any way, I am not courageous enough to join any army.

**Comment by Larpoux** [comment]
@heknt 
![d9a71bcb74f59e9fe7632fa54ba07a5a](https://user-images.githubusercontent.com/45900255/163293872-5f3787f8-c31a-460a-83d0-d79524b92f75.jpeg)

**Comment by heknt** [comment]
Hi Larpoux.

It is completely normal. We all are just making our job.
You are talking about the problem, I am paying the taxes, donating to the funds etc. That is already at least something.
We all cannot be the warriors. Therefore who will support an economy?

You can look my issue when it becomes more comfortable for you.

**Comment by Larpoux** [comment]
> It is completely normal. We all are just making our job.

I had not think like that, but you are perfectly right to point that.
I am more than 70 years old, have absolutely no knowledge for war, do not speak Russian or Ukrainian. It will not be useful for you if I go to fight in Ukraine. I am definitely not a warrior. But I feel a little guilty for that.

The only thing that I can do more or less correctly in my life is coding software. I am glad that you told me that it will help you, indirectly.

**Comment by heknt** [comment]
If you have a passion you can write the letters to your consulate and government, also twitter messages to the famous people of your country about general material or weapon support to Ukraine.

But among the all these things, I, and all my team, thank you for this unpaid work with a very nice plugin. You already did many good things during your entire life I sure.

**Comment by Larpoux** [comment]
- When I was between 20 and 30, I was too young to take correct decisions. And I was a real anti-militarian.
- Between 30 to 60, the only thing important was to stay alive for my children.
- And now that I am 70 and completely free, people say that I am too old to help in Ukraine. 
Perhaps I can join the red-cross or another organization?
(It is actually impossible to fly to Ukraine because it is closed.) 

Nikita : my private email address is larpoux@gmail.com. I would be honored to talk with you

**Comment by Larpoux** [comment]
Nikita : 
The Flutter Sound Widget UI was too buggy and not maintained.
We removed it in Flutter Sound 9.x

I suggest that you use the last version of Flutter Sound (9.2.2),
and that you do not use the Widget UI.
You must define your own UI, and call directly the FlutterSoundPlayer and FlutterSoundRecorder classes.

If you continue to have problems, you must check your asynchronous processing.
Asynchronous things are sometimes tricky, but you must be sure of your code.

```
┌───────────────────────────────────────────────────
I/flutter (28526): │ #0   FlutterSoundPlayer._cleanCompleters (package:flutter_sound_lite/public/flutter_sound_player.dart:478:15)
I/flutter (28526): │ #1   FlutterSoundPlayer.audioPlayerFinished (package:flutter_sound_lite/public/flutter_sound_player.dart:323:5)
I/flutter (28526): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (28526): │ ⚠️ Kill _stopPlayer()
I/flutter (28526): └────────────────────────────────────────────────────────
```
This happens when a playback is finished : we try to stop the player, but another `stopPlayer()` is already in progress.
Because we cannot have two `stopPlayer()` running at the same time, we kill the other pending `stopPlayer()`

Then, later this other pending `stopPlayer()` which has been killed throws an exception : 
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: killed by cleanCompleters
...
I/flutter (28526): ┌────────────────────────────
I/flutter (28526): │ #0   FlutterSoundPlayer.stopPlayerCompleted (package:flutter_sound_lite/public/flutter_sound_player.dart:438:15)
I/flutter (28526): │ #1   MethodChannelFlutterSoundPlayer.channelMethodCallHandler (package:flutter_sound_platform_interface/method_channel_flutter_sound_player.dart:137:19)
I/flutter (28526): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (28526): │ 🐛 Error : cannot process stopPlayerCompleted
I/flutter (28526): └────────────────────────────────────
...
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: killed by cleanCompleters
```

Probably the simplest for you is just ignore the exception thrown by stopPlayer().
The exception is just for unfrequent app, which need to know that the stopPlayer() failed.

If you have difficulties to understand the sequence which happens, please send again your new logs under 9.2.2.
And perhaps show me your code

**Comment by heknt** [comment]
@Larpoux thank you for an explanation. I am not using a flutter_sound Widget UI. I am using only own presentation and classes **FlutterSoundPlayer**, **FlutterSoundRecorder** from your library. I think that I fixed my problem with cracked audio player by additional checking if player is playing. If it is not playing, so I am not calling **stopPlayer** and **pausePlayer** functions.

I am not sure that I fixed the bug, I will write you if it repeats at least one more time.

And about lags - it is not really critical for me now because it produces on the rare amount of the devices according to reviews of our apps users.

Unfortunaltely, I cannot share to you our code because it is under NDA. But thank you for your help.
